float64[] final
bool initialized
bool completed
